File: backend/app/schemas/monitoring.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class ServerMetricOut(BaseModel):
    id: uuid.UUID
    server_id: uuid.UUID
    cpu_percent: float | None
    ram_percent: float | None
    ram_used_mb: int | None
    ram_total_mb: int | None
    disk_percent: float | None
    disk_used_gb: float | None
    disk_total_gb: float | None
    load_1: float | None
    load_5: float | None
    load_15: float | None
    uptime_seconds: int | None
    recorded_at: datetime

    model_config = {"from_attributes": True}


class AlertCreate(BaseModel):
    metric: str
    condition: str
    threshold: float
    channel: str
    channel_target: str


class AlertOut(BaseModel):
    id: uuid.UUID
    server_id: uuid.UUID
    metric: str | None
    condition: str | None
    threshold: float | None
    channel: str | None
    channel_target: str | None
    is_active: bool
    last_triggered: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/monitoring.ts
import { apiClient } from "./client"
import type { Alert } from "@/types"

// ── Types ──────────────────────────────────────────────────────────────────

export interface MetricPoint {
  id: string
  server_id: string
  cpu_percent: number | null
  ram_percent: number | null
  ram_used_mb: number | null
  ram_total_mb: number | null
  disk_percent: number | null
  disk_used_gb: number | null
  disk_total_gb: number | null
  load_1: number | null
  load_5: number | null
  load_15: number | null
  uptime_seconds: number | null
  recorded_at: string
}

export interface AlertCreateBody {
  metric: "cpu" | "ram" | "disk"
  condition: "gt" | "gte" | "lt" | "lte"
  threshold: number
  channel: "email" | "webhook" | "slack"
  channel_target: string
}

// ── Metrics ────────────────────────────────────────────────────────────────

export async function getMetricsHistory(
  serverId: string,
  hours = 24
): Promise<MetricPoint[]> {
  const res = await apiClient.get<MetricPoint[]>(
    `/api/servers/${serverId}/metrics/history`,
    { params: { hours } }
  )
  return res.data
}

// ── Alerts ─────────────────────────────────────────────────────────────────

export async function listAlerts(serverId: string): Promise<Alert[]> {
  const res = await apiClient.get<Alert[]>(`/api/servers/${serverId}/alerts`)
  return res.data
}

export async function createAlert(
  serverId: string,
  body: AlertCreateBody
): Promise<Alert> {
  const res = await apiClient.post<Alert>(`/api/servers/${serverId}/alerts`, body)
  return res.data
}

export async function updateAlert(
  alertId: string,
  body: AlertCreateBody
): Promise<Alert> {
  const res = await apiClient.put<Alert>(`/api/alerts/${alertId}`, body)
  return res.data
}

export async function deleteAlert(alertId: string): Promise<void> {
  await apiClient.delete(`/api/alerts/${alertId}`)
}

export async function toggleAlert(alertId: string): Promise<Alert> {
  const res = await apiClient.post<Alert>(`/api/alerts/${alertId}/toggle`)
  return res.data
}

export async function testAlert(alertId: string): Promise<void> {
  await apiClient.post(`/api/alerts/${alertId}/test`)
}
